feat: add argument types parsing

ifedchankau opened the pull request:
- Add parsing of arguments types
- Flags now have type:null
- Add dictionary for argument types
- Add tests for argument types

Closes #12

romanrostislavovich commented on src/template/context.js:
```diff
@@ -43,6 +43,31 @@ const context = {
+            'integer',
+            'number',
+        ],
+        array: [
```
We don't have a `array` type

romanrostislavovich replied:
I'm not sure, but better put the `argumenTypes` in separate file? Because we use it only for identity types of arguments

romanrostislavovich commented on src/arguments.js:
```diff
@@ -89,18 +92,29 @@ const isValueBoolean = (string) => {
-    let isType = false;
+const getPropertyType = (string, context) => {
+    let isFlag = true;
+    let argumentType = 'default';
```
Possible do you mean a `null` value in argumentType ?

ifedchankau replied:
@svoboda-rabstvo no, `null` have flags, and `default` will have all arguments if no types was found. `default` will be replaced later for value from `template/option.json`.

romanrostislavovich replied:
I think it's no need. We can create implementation without it and it's complicated the implementation, also mislead users

ifedchankau replied:
@svoboda-rabstvo done.

romanrostislavovich commented on src/arguments.js:
```diff
@@ -89,18 +92,29 @@ const isValueBoolean = (string) => {
-        isType = argumentSubstring.indexOf('-') !== 0 ? true : isType;
+        isFlag = argumentSubstring.indexOf('-') !== 0 ? false : isFlag;
+        if (!isFlag) {
+            Object.keys(context.argumentTypes).forEach((type) => {
```
I think we need to create a separate function which identity types of argument, @Arcle1 what do you think?

ifedchankau replied:
@svoboda-rabstvo done. Please, recheck PR.

romanrostislavovich commented on test/parsing/files/delimiter.equal.valid.json:
```diff
@@ -14,7 +14,7 @@
             "--ignore": {
                "id": "--ignore",
-               "type": "string",
+               "type": "null",
```
`null` is useg without quotations

ifedchankau replied:
In our EP (https://github.com/repometric/ep/blob/master/ep/lh/ep-lh-0003.md) we have `"type": "null"`. Is it mistake in EP?

romanrostislavovich replied:
Yes

ifedchankau replied:
@svoboda-rabstvo i fixed it. Please, recheck PR.

romanrostislavovich commented on src/templatizer.js:
```diff
@@ -17,25 +17,29 @@ const templatizer = (context, config) => {
                 break;
             case '':
                 optionSchema.id = customPrefix + 'path';
                 option.description = 'Path to file or folder to analyze';
```
I think we can change `option` to `optionSchema` here

romanrostislavovich commented on src/templatizer.js:
```diff
@@ -17,25 +17,29 @@ const templatizer = (context, config) => {
                 break;
             case '--stdin':
                 optionSchema.id = customPrefix + 'stdin';
+                option.type = 'string';
```
Why do you use `option.type`, but not `optionSchema.type` ?

ifedchankau replied:
@svoboda-rabstvo because it's assigning later for **all** arguments. But, yes, maybe it is better to do it at once. I will change it.

ifedchankau committed:
feat: add argument types parsing

- Add parsing of arguments types
- Flags now have type:null
- Add dictionary for argument types
- Add tests for argument types

Closes #12

romanrostislavovich committed:
Merge branch 'develop' into feat/argument-types

romanrostislavovich reviewed (approved):
+1

romanrostislavovich commented:
:tada: This PR is included in version 1.0.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@linterhub/usage-parser)
- [GitHub release](https://github.com/linterhub/usage-parser/releases/tag/v1.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
